Adicionar nova solução ao exercício 02_02

PmaiWoW opened the pull request.

NunoCarrico98 commented:
Feito!

fakenmc commented:
Obrigado Nuno, solução aceite.

PmaiWoW committed:
Remover solução de 01_18 adicionada por erro

PmaiWoW committed:
Merge remote-tracking branch 'upstream/master'

PmaiWoW committed:
Merge remote-tracking branch 'upstream/master'

PmaiWoW committed:
Merge remote-tracking branch 'upstream/master'

PmaiWoW committed:
Adicionar nova solução ao exercício 02 de 02(Matéria de LP2)

nunofachada reviewed (changes requested):
* A tua solução é a 2, não a 1.
* Na classe `Squad`, teres uma lista pública - mesmo sob a forma de uma propriedade - é má ideia, pois permites que código externo faça o que quiser à lista, nomeadamente apagar tudo o que está lá dentro. A lista é um detalhe de implementação de `Squad` que não deve ser exposto a terceiros. Se no futuro decidires usar uma `Stack` ou um `HashSet` em vez de uma `List` colocas em risco todo o código que assumia que antes tinha uma lista. A funcionalidade de adicionar e remover unidades deve ser exposta com métodos próprios em `Squad`, e nesses métodos sim, colocas e removes coisas dentro da lista, _stack_, _queue_ ou o que for.
* Fará sentido teres dois construtores em `Squad`? Repara que na prática ambos aceitam um `IEnumerable<IUnit>`! No caso do construtor com `params`, achas que algumas vez a nível de código do jogo vais querer colocar duas, três, cinco unidades explicitamente?
* Para que precisas do `private set` nas propriedades `UnitName`, `Position` e `Health` se nunca o usas? Já agora, experimenta lá usar isso uma vez (tipo num construtor) só para veres o que acontece...

PmaiWoW committed:
Corrigir o acesso externo à lista contida na classe Squad;
Corrigir tipos de coleções usados na classe Squad, tornando a funcionalidade generalizada;
Corrigir construtor da classe Squad e eliminar o construtor desnecessário;
Eliminar private set nas propriedades UnitName, Position e Health da classe Squad;

nunofachada reviewed (changes requested):
* `ICollection<T>` oferece os métodos `Add()`, `Remove()` e `Clear()`, permitindo ao código externo alterar a coleção interna de `Squad`.
* As propriedades só de leitura só com uma linha e um lambda podem ser ainda mais simplificadas (é mesmo preciso estar lá o `get`?).
* No construtor de `Squad`, podias usar [este](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.-ctor?view=netframework-4.8#System_Collections_Generic_List_1__ctor_System_Collections_Generic_IEnumerable__0__) construtor de `List<T>` em vez de fazeres um `foreach`.
* No método `Main()` as variáveis `unitBundle1` e `unitBundle2` não podem ser mais gerais? Fazes alguma coisa com elas que exija os métodos definidos em `ICollection<T>`?

PmaiWoW committed:
Corrigir vários aspetos, de acordo com o feedback:
Corrigir tipo da coleção interna da classe Squad de ICollection<T> para IEnumerable<T>;
Simplificar propriedades só de leitura para caberem numa linha usando lambdas;
Alterar construtor da classe Squad, usando de um construtor de List<T> existente;
Alterar tipo das variáveis unitBundle1 e unitBundle2 para IEnumerable<T>, generalizando-as;

nunofachada reviewed (changes requested):
* Falta um `using` na classe `Program`.
* No teste que fazes no `Main()` podias adicionar uns `WriteLines()` pelo meio a dizer o que estás a fazer e a que corresponde cada uma das coisas impressas. Percebia-se melhor.

PmaiWoW committed:
Corrigir partes da resposta de acordo com o feedback:
Adicionar using necessário à classe Program;
Adicionar WriteLines() que descrevem ao utilizador o que está a acontecer no ecrã;

nunofachada commented:
Obrigado Pedro, solução aceite.
